File: Tema2/src/python/main.py
import sys
import time
import tracemalloc

def get_collatz_steps(n):
    steps = 0
    val = n
    while val > 1:
        if val % 2 == 0:
            val = val // 2
        else:
            val = 3 * val + 1
        steps += 1
    return steps

def run_benchmark(limit):
    tracemalloc.start()
    start_time = time.perf_counter()
    
    max_steps = 0
    max_num = 0
    
    for i in range(1, limit + 1):
        steps = get_collatz_steps(i)
        if steps > max_steps:
            max_steps = steps
            max_num = i
            
    end_time = time.perf_counter()
    current, peak = tracemalloc.get_traced_memory()
    tracemalloc.stop()
    
    elapsed_ms = (end_time - start_time) * 1000.0
    peak_mb = peak / (1024 * 1024)
    
    print(f"Max steps: {max_steps} for number: {max_num}")
    print(f"Time: {elapsed_ms:.2f} ms")
    print(f"Memory: {peak_mb:.4f} MB")

if __name__ == "__main__":
    limit = 2000000
    if len(sys.argv) > 1:
        try:
            limit = int(sys.argv[1])
        except ValueError:
            pass
    run_benchmark(limit)

File: Tema2/src/javascript/main.js
const { performance } = require('perf_hooks');

function getCollatzSteps(n) {
    let steps = 0;
    let val = n;
    while (val > 1) {
        if (val % 2 === 0) {
            val = val / 2;
        } else {
            val = 3 * val + 1;
        }
        steps++;
    }
    return steps;
}

function runBenchmark(limit) {
    // Warm up and gc if possible (node flags)
    const startMemory = process.memoryUsage().heapUsed;
    const startTime = performance.now();
    
    let maxSteps = 0;
    let maxNum = 0;
    
    for (let i = 1; i <= limit; i++) {
        const steps = getCollatzSteps(i);
        if (steps > maxSteps) {
            maxSteps = steps;
            maxNum = i;
        }
    }
    
    const endTime = performance.now();
    const endMemory = process.memoryUsage().heapUsed;
    
    const elapsedMs = endTime - startTime;
    // Estimate memory overhead during execution
    const peakMemoryMb = (process.memoryUsage().rss) / (1024 * 1024);
    
    console.log(`Max steps: ${maxSteps} for number: ${maxNum}`);
    console.log(`Time: ${elapsedMs.toFixed(2)} ms`);
    console.log(`Memory: ${peakMemoryMb.toFixed(4)} MB`);
}

function main() {
    let limit = 2000000;
    if (process.argv.length > 2) {
        const argLimit = parseInt(process.argv[2], 10);
        if (!isNaN(argLimit)) {
            limit = argLimit;
        }
    }
    runBenchmark(limit);
}

main();
